control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java

```diff
@@ -35,58 +39,62 @@
     public void handlePost(ActionParameters params) throws ActionException {
         params.requireLoggedInUser();
 
-        JSONObject jsonPayload = params.getHttpParamAsJSON("featureData");
-        OskariLayer layer = getLayer(jsonPayload.optString("layerId"));
-
-        if (!canEdit(layer, params.getUser())) {
-            throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
-        }
+        try {
+            JSONArray paramFeatures = new JSONArray(params.getHttpParam("featureData"));
+            JSONArray updatedFeatureIds = new JSONArray();
+            for (int i = 0; i < paramFeatures.length(); i++) {
+                JSONObject featureJSON = paramFeatures.getJSONObject(i);
+                OskariLayer layer = getLayer(featureJSON.optString("layerId"));
+
+                if (!canEdit(layer, params.getUser())) {
+                    throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
+                }
 
-        final WFSLayerConfiguration wfsMetadata = getWFSConfiguration(layer.getId());
-        final String wfstMessage = createWFSTMessage(jsonPayload, layer, wfsMetadata);
-        LOG.debug("Inserting feature to service at", layer.getUrl(), "with payload", wfstMessage);
-        final String responseString = postPayload(layer, wfstMessage);
-        final String updatedFeatureId = parseFeatureIdFromResponse(responseString);
+                final String wfstMessage = createWFSTMessage(featureJSON);
+                LOG.debug("Inserting feature to service at", layer.getUrl(), "with payload", wfstMessage);
+                final String responseString = postPayload(layer, wfstMessage);
+                updatedFeatureIds.put(parseFeatureIdFromResponse(responseString));
 
-        flushLayerTilesCache(layer.getId());
-        ResponseHelper.writeResponse(params, JSONHelper.createJSONObject("fid", updatedFeatureId));
+                flushLayerTilesCache(layer.getId());
```

Comment:
This could be improved by collecting all distinct layerIds updated in the loop in to a Set and then flushing the layers after this loops is finished. That way we only flush each different layer once.